libs/cart/state/src/effects/cart-coupon.effects.ts

```diff
@@ -63,10 +64,19 @@ export class DaffCartCouponEffects<
       map(() => this.storage.getCartId()),
       switchMap(cartId => this.driver.apply(cartId, action.payload)),
       map(resp => new DaffCartCouponApplySuccess(resp)),
-      catchError(error => of(error instanceof DaffStorageServiceError
-        ? new DaffCartStorageFailure(this.errorMatcher(error))
-        : new DaffCartCouponApplyFailure(this.errorMatcher(error)),
-      )),
+      catchError(error => {
+        switch (true) {
+          case error instanceof DaffStorageServiceError:
+            return of(new DaffCartStorageFailure(this.errorMatcher(error)));
+
+          case error instanceof DaffInvalidCouponCodeError:
+            (<DaffInvalidCouponCodeError>error).coupon = action.payload.code;
```

Comment:
Shouldn't this be handled where the error is thrown?

Reply:
The implementation is non-trivial. The code isn't in the error and the error transformer would have to be rewritten to accomodate that special case.

Reply:
What precisely does the relevant Magento error response look like?

Reply:
```
extensions: {category: "graphql-no-such-entity"}
locations: Array(1)
0: {line: 1, column: 65}
length: 1
__proto__: Array(0)
message: "The coupon code isn't valid. Verify the code and try again."
path: ["applyCouponToCart"]
```

Reply:
The more I look at this, the more I think this shouldn't be in `@daffodil/cart/state`. Because the core code of the daffodil packages (agnostic to platform) shouldn't have platform specific code in it, and if the error message we're given from Magento doesn't have the coupon code, then the frontend shouldn't be responsible for knowing which coupon code caused the error. But I think the only other option is to write something custom on the backend, right?

Reply:
> and if the error message we're given from Magento doesn't have the coupon code, then the frontend shouldn't be responsible for knowing which coupon code caused the error

But that is connecting the platform-agnostic frontend to the specific platform.

> Because the core code of the daffodil packages (agnostic to platform) shouldn't have platform specific code in it

What part of the code is platform-specific?

Reply:
> But I think the only other option is to write something custom on the backend, right?

The error code in the Magento driver can be reworked to set the coupon code.

Reply:
> What part of the code is platform-specific?

There is nothing explicitly platform-specific, but it is implicitly platform specific (meaning there is nothing named "magento", but there is code here that only exists because of the specific way that magento works). For example, if we implement Shopify and they do give the coupon code in the error message. That would make this code unnecessary for the shopify implementation, and it would only exist because magento doesn't give the coupon code in the error message. In that case the code that adds the coupon code to the daffodil error would be platform specific. If we could show that no platforms return the coupon code in the error messages, then it could be said that this is not platform specific. Unfortunately, there isn't a unifying set of rules that all platforms follow, so we can't know this and must assume that it is specific to magento.

> But that is connecting the platform-agnostic frontend to the specific platform's behavior.

I'm not sure what you mean by "that" here. As I've argued above, I think this code itself makes the frontend code, not platform-agnostic.

> The error transformer in the Magento driver can be reworked to set the coupon code.

If this works, that's cool. So would you need to send the request payload into the error transformer along with the magento error?

Reply:
Oh I think I see what you meant by this:

>>and if the error message we're given from Magento doesn't have the coupon code, then the frontend shouldn't be responsible for knowing which coupon code caused the error.
>But that is connecting the platform-agnostic frontend to the specific platform's behavior.

I worded that kind of strangely. What I should have said is "the frontend shouldn't be responsible for knowing which coupon code caused the error, even if the error message we're given from Magento doesn't have the coupon code. It is magento's responsibility to pass @daffodil/state an error message with the coupon code."

Reply:
> So would you need to send the request payload into the error transformer along with the magento error?

Yes

> there is code here that only exists because of the specific way that magento works

That makes sense.

> I'm not sure what you mean by "that" here. As I've argued above, I think this code itself makes the frontend code, not platform-agnostic.

Well I think this is tangential to the main point about where to set the coupon code on the error but:

> if the error message we're given from Magento doesn't have the coupon code

so here you reference a platform-specific behavior

> then the frontend shouldn't be responsible for knowing which coupon code caused the error

and then make a conclusion about platform-agnostic frontend behavior. So in that sense I say you have connected them.

I think its okay to draw inspiration for our agnostic frontend models from specific platforms but I think its wrong to create a hard link between their behaviors in this way. I think its reasonable for a frontend to expect a coupon error to contain information about which code caused the error; I see this on many ecommerce sites.

Back to the main point, I also think its reasonable to demand that the driver layer set the coupon code on the error despite how ugly that implementation might be. That's the purpose of our drivers: to encapsulate all the quirks and shortcomings of each individual platform behind a single well-defined interface. And although the interface doesn't specify the errors that can be thrown (and maybe it should), I agree that a fully-formed invalid coupon code error is implicitly part of the cart coupon driver's behavior.

I'll go ahead and make this change

Reply:
> I worded that kind of strangely. What I should have said is "the frontend shouldn't be responsible for knowing which coupon code caused the error, even if the error message we're given from Magento doesn't have the coupon code. It is the platform's responsibility to pass @daffodil/state an error message with the coupon code."

okay I understand what you mean. yeah I briefly considered leaving it to the app to figure out which code caused the error but I came to the same conclusion: the app, and to a certain extend the state layer, isn't responsible for those kinds of things. It really is the responsibility of the backend, and therefore the drivers (if the backend doesn't do it), to provide that info.

Reply:
@lderrickable @damienwebdev changed

Reply:
Very much agree with this reasoning.